smithy-model/src/main/java/software/amazon/smithy/model/knowledge/IdentifierBindingIndex.java

```diff
@@ -87,22 +93,73 @@ public BindingType getOperationBindingType(ToShapeId resource, ToShapeId operati
+     * value for that identifier.
+     *
+     * @param resource Shape ID of a resource.
+     * @param operation Shape ID of an operation.
+     * @return Returns the identifier bindings map or an empty map if the
+     *  binding is invalid or cannot be found.
+     */
+    public Map<String, String> getOperationOutputBindings(ToShapeId resource, ToShapeId operation) {
+        return Optional.ofNullable(outputBindings.get(resource.toShapeId()))
                 .flatMap(resourceMap -> Optional.ofNullable(resourceMap.get(operation.toShapeId())))
                 .map(Collections::unmodifiableMap)
                 .orElseGet(Collections::emptyMap);
     }
 
+    /**
+     * Gets a map of identifier names to input member names that provide a
+     * value for that identifier.
+     *
+     * @deprecated Use {@link #getOperationInputBindings} instead.
+     *
+     * @param resource Shape ID of a resource.
+     * @param operation Shape ID of an operation.
+     * @return Returns the identifier bindings map or an empty map if the
+     *  binding is invalid or cannot be found.
+     */
+    @Deprecated
+    public Map<String, String> getOperationBindings(ToShapeId resource, ToShapeId operation) {
+        return getOperationInputBindings(resource, operation);
+    }
+
+    /**
+     * Gets a set of identifier names associated with either input or output on the operation.
+     *
+     * @param resource Shape ID of a resource.
+     * @param operation Shape ID of an operation.
+     * @return Returns a set of all identifier names associated with the operation.
+     */
+    public Set<String> getAllOperationIdentifiers(ToShapeId resource, ToShapeId operation) {
+        return SetUtils.copyOf(allIdentifiers);
+    }
+
     private void processResource(ResourceShape resource, OperationIndex operationIndex, Model model) {
-        bindings.put(resource.getId(), new HashMap<>());
+        inputBindings.put(resource.getId(), new HashMap<>());
+        outputBindings.put(resource.getId(), new HashMap<>());
         bindingTypes.put(resource.getId(), new HashMap<>());
         resource.getAllOperations().forEach(operationId -> {
             // Ignore broken models in this index.
-            Map<String, String> computedBindings = operationIndex.getInputShape(operationId)
+            Map<String, String> computedInputBindings = operationIndex.getInputShape(operationId)
                     .map(inputShape -> computeBindings(resource, inputShape))
                     .orElseGet(HashMap::new);
-            bindings.get(resource.getId()).put(operationId, computedBindings);
+            inputBindings.get(resource.getId()).put(operationId, computedInputBindings);
+            allIdentifiers.addAll(computedInputBindings.keySet());
+
+            Map<String, String> computedOutputBindings = operationIndex.getOutputShape(operationId)
+                    .map(outputShape -> computeBindings(resource, outputShape))
+                    .orElseGet(HashMap::new);
```

Comment:
Is the map immutable at this point? If so, then you could use `orElse(Collections.emptyMap())` and avoid the extra hashmap.